localparticipant.go

```diff
@@ -269,42 +269,107 @@ func (p *LocalParticipant) closeTracks() {
 	}
 }
 
-func (p *LocalParticipant) PublishData(
-	payload []byte,
-	opts ...DataPublishOption,
-) error {
+func (p *LocalParticipant) publishData(kind livekit.DataPacket_Kind, dataPacket *livekit.DataPacket) error {
+	if err := p.engine.ensurePublisherConnected(true); err != nil {
+		return err
+	}
+
+	encoded, err := proto.Marshal(dataPacket)
+	if err != nil {
+		return err
+	}
+
+	return p.engine.GetDataChannel(kind).Send(encoded)
+}
+
+// PublishData sends custom user data via WebRTC data channel.
+//
+// By default, the message can be received by all participants in a room,
+// see WithDataPublishDestination for choosing specific participants.
+//
+// Messages are sent via a LOSSY channel by default, see WithDataPublishReliable for sending reliable data.
+//
+// Deprecated: Use PublishDataPacket with UserData instead. Note that it sends reliable packets by default.
+func (p *LocalParticipant) PublishData(payload []byte, opts ...DataPublishOption) error {
 	options := &dataPublishOptions{}
 	for _, opt := range opts {
 		opt(options)
 	}
-	packet := &livekit.UserPacket{
-		Payload:               payload,
-		DestinationIdentities: options.DestinationIdentities,
+	if options.Reliable == nil {
+		// Old logic sends packets as lossy by default.
+		opts = append(opts, WithDataPublishReliable(false))
 	}
-	if options.Topic != "" {
-		packet.Topic = proto.String(options.Topic)
+	return p.PublishDataPacket(UserData(payload), opts...)
+}
+
+type DataPacket interface {
+	ToProto() *livekit.DataPacket
+}
+
+var (
+	_ DataPacket = UserData(nil)
+	_ DataPacket = (*UserDataPacket)(nil)
+	_ DataPacket = (*livekit.SipDTMF)(nil) // implemented in the protocol package
```

Comment:
Could you comment on why these definitions are needed?

Reply:
Done